internal/ask/form/aggregation.go

```diff
@@ -0,0 +1,374 @@
+	groupedSubmissions, err := GroupSubmissions(context, db, id, 0, 0, submission.SearchOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a container for the grouped aggregations.
+	groupAggregations := make(map[string]Aggregation)
+
+	// Loop through the groups of submissions.
+	for group, submissions := range groupedSubmissions {
+
+		// Perform the multiple choice aggregations.
+		mcAggregations, err := MCAggregate(context, db, id, submissions)
+		if err != nil {
+			return nil, err
+		}
+
+		// Including the aggregation of the group that is being aggregated is redundant, remove.
+		for key, agg := range mcAggregations {
+			if agg.Question == group.Question {
+				delete(mcAggregations, key)
+			}
+		}
+
+		// Pack them in an aggregation along with the submission count and group.
+		agg := Aggregation{
+			Group: group,
+			Count: len(submissions),
+			MC:    mcAggregations,
+		}
+
+		// If this is the "all" group, set the key to all as it is not an answer.
+		groupKey := ""
+		if group.Question == "all" {
+			groupKey = "all"
+		}
+
+		// Groups are ultimately based on a chosen answer. We do not have any keys for answers
+		// so hash the answer text for a unique key.
+		if group.Question != "all" {
+			hasher := md5.New()
+			hasher.Write([]byte(group.Answer))
+			groupKey = hex.EncodeToString(hasher.Sum(nil))
+		}
+
+		// Add the aggregation to the map.
+		groupAggregations[groupKey] = agg
+
+	}
+
+	return groupAggregations, nil
+}
+
+//==============================================================================
+
+// SubmissionGroup is a transport that defines the transport structure for a submission group.
+type SubmissionGroup struct {
+	Submissions map[Group][]submission.Submission `json:"submissions" bson:"submissions"`
+}
+
+// GroupSubmissions organizes submissions by Group. It looks for questions with the group by flag
+// and creates Group structs.
+func GroupSubmissions(context interface{}, db *db.DB, id string, limit int, skip int, opts submission.SearchOpts) (map[Group][]submission.Submission, error) {
+
+	if !bson.IsObjectIdHex(id) {
+		log.Error(context, "TextAggregate", ErrInvalidID, "Completed")
+		return nil, ErrInvalidID
+	}
+
+	// Get the submissions for the form.Collection
+	subs, err := submission.Search(context, db, id, limit, skip, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[Group][]submission.Submission)
+
+	// Scan all the submissions and answers.
+	for _, sub := range subs.Submissions {
+
+		// Add all submissions to the [all,all] group
+		group := Group{
+			ID:       "all",
+			Question: "all",
+			Answer:   "all",
+		}
+		tmp := groups[group]
+		tmp = append(tmp, sub)
+		groups[group] = tmp
+
+		// Look for answers that may contain sub groups.
+		for _, ans := range sub.Answers {
+
+			// Skip nil answers.
+			if ans.Answer == nil {
+				continue
+			}
+
+			// Only include answers of questions flagged with "includeInGroups".
+			props := ans.Props.(bson.M)
```

Comment:
It nees to check if is really bson.M as it seems that sometimes ans.Props could be nil. 

props, ok := ans.Props.(bson.M)